rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+*   We still cannot expose file-based or record-based interface to end-users in
+    the TF trainer. As that is a direct result of I/O + parsing being abstracted
+    out.
+
+The disadvantages of this approach are:
+
+*   This tightly couples TF with TFX components, in the following ways:
+
+    *   Components will need TF to read the data in.
+    *   Components that analyze the data (e.g. TFDV) will operate against
+        StructuredTensors, and the easiest way to to conduct certain
+        computations (for example, slicing a StructuredTensor or computing the
+        mean) with StructuredTensors is through TF ops and their python
+        bindings.
+
+    Such a coupling is not in our favor because:
+
+    *   Some TFX components functionally do not require TF to work. For example,
+        TFDV can analyze any data set. TFMA can analyze a model while treating
+        the model as a blackbox. In both cases, the ML library that trained the
+        model is irrelevant and TF should not be assumed. And coupling with TF
+        does not only introduce a heavy dependency, but also forces the user to
+        learn about TF if they need to implement a TFXIO for their data format.
+
+    *   Operations against StructuredTensors boils down to Python + TF ops. The
+        overhead of either is much higher than just calling an Arrow C++ API
+        that does the same operation.
+
+    *   The main extension point of TF is Ops which don't understand the nested
+        structures. Compared to using Arrow’s C++ APIs, implementing an Op that
+        deals with StructuredTensors will be much complicated.
+
+# Questions and Discussion Topics
+
+## OSS build / release issues
+
+### We don’t have a strong representation in the Arrow community
+
+This has led to some issues. For example, the PyPI/Wheel packaging for pyarrow
+currently is unfunded and lacks volunteers and there is a risk of support being
+dropped, but we do rely on the pyarrow wheel as TFX is released on PyPI.
+
+### ABI compatibility with libarrow
+
+The OSS library, tfx_bsl will depend on Arrow and TensorFlow’s DSOs (dynamic
+shared objects). Because both libraries currently expose C++ APIs, there are
+always risks of incompatible ABIs as TensorFlow and Arrow are likely to be built
+using different toolchains, we cannot completely eliminate the risks.
+
+With [Modular TensorFlow](https://github.com/tensorflow/community/pull/77),
+which replaced all the C++ APIs with C-APIs, we will be able to eliminate the
+risk by using the same toolchain that builds Arrow.
+
+Furthermore, the Apache Arrow community is discussing about an
+[ABI-stable C-struct](https://github.com/apache/arrow/pull/5442) that describes
+Arrow Arrays. This will allow to build Apache Arrow from source and link
+statically with our code, and only talk with pyarrow through that ABI-stable
+interface.
+
+Since in Google we build everything from HEAD, using the same toolchain, there
```

Comment:
But we need to support opensource too, so ABI issues are nontrivial.

Reply:
So far the ABI issue has been manageable: tfx_bsl is already built with libarrow.so dynamically linked, and the problem can be greatly mitigated with https://github.com/apache/arrow/pull/5442, which would allow tfx_bsl build statically with arrow source and talk with pyarrow through this ABI-stable interface.